### ISSUE TITLE
Calling a mutation within React.useEffect causes infinite loop

### ISSUE THREAD

**Issue by jonlambert** [trojanowski/react-apollo-hooks#205]
I'm attempting to perform a mutation when a specific route is hit in my application. When using useEffect and passing through a mutation method as a dependancy, it enters an infinite loop.``

```javascript
const [myMutation] = useMutation(MY_MUTATION);
const Component = () => {
  React.useEffect(() => {
        myMutation();
    }, [myMutation]);
}
```

This may be by design and my code an anti-pattern, if so I'd love to know an alternative way to trigger a mutation when a component is mounted for the first time.

A similar issue was fixed in [`graphql-hooks`](https://github.com/nearform/graphql-hooks/issues/234) recently.

**Comment by movahedan** [comment]
I seems it enters an infinite loop because of "myMutation" in the useEffect second argument. It happens because every time that a mutation calls, returned function "myMutation" also changes. Did you try it without that argument ? I know it causes a warning, but maybe it'll fix your problem.

**Comment by jonlambert** [comment]
Yeah whilst that does fix the problem, unfortunately that goes against the official spec for useEffect and breaks hook linting. I'd like to find a way to fix this without having to disable the linter for that line.

**Comment by cglacet** [comment]
Did you found a way to solve this issue? Is there a reason for `myMutation` to change every time it's called? I agree that this looks strange to omit it from the dependency list, I would also like to avoid this.

**Comment by yaralahruthik** [comment]
Yes, your suggestions do fix the problem but give the lint warning. But the proper way of doing this is to destructure the mutate from the mutation and provide that to the dependency array of use effect!
`const {mutate} = useMutation(someFunction)`

useEffect(() => {
  mutate()
}, [mutate])

**Comment by Oracard** [comment]
Omg thank you @yaralahruthik! This solved my identical issue perfectly without requiring weird workarounds :)

**Comment by janusqa** [comment]
I have a similar setup and I use 
```js
const { mutate: renameMutateForReasons} = useMutation(MyMutation);

const Component = () => {
  React.useEffect(() => {
        renameMutateForReasons();
    }, [renameMutateForReasons]);
}
```

I don't get the repetitions.

**Comment by ahmedcynoteck** [comment]
I am also getting an infinite loop and i am using useMutation  from '@tanstack/react-query';

  const {
    mutate,
    data: fetchedUsers,
    isLoading: isLoadingUsers,
    error,
    isError: isLoadingUsersError,
  }: any = useMutation({
    mutationFn: () => useUserMutation(USERS_KEY, countyid, token),
    mutationKey: ['userslist'],
    retry: false,
  });

  useEffect(() => {
    mutate();
  }, []);

**Comment by yaralahruthik** [comment]
> I am also getting an infinite loop and i am using useMutation from '@tanstack/react-query';
> 
> const { mutate, data: fetchedUsers, isLoading: isLoadingUsers, error, isError: isLoadingUsersError, }: any = useMutation({ mutationFn: () => useUserMutation(USERS_KEY, countyid, token), mutationKey: ['userslist'], retry: false, });
> 
> useEffect(() => { mutate(); }, []);

In your case the issue does not seem to be in this part of the code. Is it possible that the mutation updates some data which results in re-rendering the component that calls it? Maybe try commenting out parts of code to figure the actual cause?

**Comment by alvesvin** [comment]
Weird API. You end up having to rename mutate and mutateAsync every time in order to get this right.

**Comment by dominictobias-bullish** [comment]
> Weird API. You end up having to rename mutate and mutateAsync every time in order to get this right.

Agree this is pretty weird and breaks the normal interior mutability pattern (e.g. seen in `useRef`) where you can rely on the outer object being constant, or `useMemo` where it's constant unless something changes

**Comment by Kais3rP** [comment]
Agree, weird behaviour, stumbled on it today...
